manager/shard_strategy.go

```diff
@@ -0,0 +1,191 @@
+		}
+
+		return &HashShardStrategy{
+			enableUncoveredReplica: shardConfig.EnableUncoveredReplica,
+			podCount:               shardConfig.PodCount,
+		}, nil
+	case config.ProjectShardType, config.DomainShardType:
+		replicas := make([][]string, 0)
+		for _, replica := range shardConfig.Replicas {
+			if len(replica.Entities) == 0 {
+				return nil, fmt.Errorf("unable to create replica with 0 configured entity(ies)")
+			}
+
+			replicas = append(replicas, replica.Entities)
+		}
+
+		var envType environmentType
+		switch shardConfig.Type {
+		case config.ProjectShardType:
+			envType = project
+		case config.DomainShardType:
+			envType = domain
+		}
+
+		return &EnvironmentShardStrategy{
+			enableUncoveredReplica: shardConfig.EnableUncoveredReplica,
+			envType:                envType,
+			replicas:               replicas,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("shard strategy '%s' does not exist", shardConfig.Type)
+}
+
+// HashShardStrategy evenly assigns disjoint keyspace responsibilities over a collection of pods.
+// All FlyteWorkflows are assigned a shard-key using a hash of their executionID and are then
+// processed by the FlytePropeller instance responsible for that keyspace range.
+type HashShardStrategy struct {
+	enableUncoveredReplica bool
+	podCount               int
+}
+
+func (h *HashShardStrategy) GetPodCount() int {
+	if h.enableUncoveredReplica {
+		return h.podCount + 1
+	}
+
+	return h.podCount
+}
+
+func (h *HashShardStrategy) UpdatePodSpec(pod *v1.PodSpec, podIndex int) error {
+	container, err := getFlytePropellerContainer(pod)
+	if err != nil {
+		return err
+	}
+
+	if podIndex >= 0 && podIndex < h.podCount {
+		startKey, endKey := computeKeyRange(v1alpha1.ShardKeyspaceSize, h.podCount, podIndex)
+		for i := startKey; i < endKey; i++ {
+			container.Args = append(container.Args, "--propeller.include-shard-label", fmt.Sprintf("%d", i))
+		}
+	} else if h.enableUncoveredReplica && podIndex == h.podCount {
+		for i := 0; i < v1alpha1.ShardKeyspaceSize; i++ {
+			container.Args = append(container.Args, "--propeller.exclude-shard-label", fmt.Sprintf("%d", i))
+		}
+	} else {
+		return fmt.Errorf("invalid podIndex '%d' out of range [0,%d)", podIndex, h.GetPodCount())
+	}
+
+	return nil
+}
+
+// Computes a [startKey, endKey) pair denoting the key responsibilities for the provided pod index
+// given the keyspaceSize and podCount parameters
+func computeKeyRange(keyspaceSize, podCount, podIndex int) (int, int) {
+	keysPerPod := keyspaceSize / podCount
+	keyRemainder := keyspaceSize - (podCount * keysPerPod)
+
+	return computeStartKey(keysPerPod, keyRemainder, podIndex), computeStartKey(keysPerPod, keyRemainder, podIndex+1)
+}
+
+func computeStartKey(keysPerPod, keysRemainder, podIndex int) int {
+	return (intMin(podIndex, keysRemainder) * (keysPerPod + 1)) + (intMax(0, podIndex-keysRemainder) * keysPerPod)
+}
+
+func intMin(a, b int) int {
```

Comment:
i know sad, we should just write these in flytestdlib one day

Reply:
+1